File: android/src/main/java/com/lightningkite/kwift/views/LinearLayout.actual.kt
package com.lightningkite.kwift.views

import android.view.Gravity
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import android.widget.LinearLayout
import com.lightningkite.kwift.views.geometry.Align
import com.lightningkite.kwift.views.geometry.AlignPair

fun LinearLayout.params(
    sizeX: Int = 0,
    sizeY: Int = 0,
    marginStart: Int = 0,
    marginEnd: Int = 0,
    marginTop: Int = 0,
    marginBottom: Int = 0,
    gravity: AlignPair = AlignPair.center,
    weight: Float = 0f
): LinearLayout.LayoutParams {
    val d = resources.displayMetrics.density
    val align = if(orientation == LinearLayout.HORIZONTAL) gravity.vertical else gravity.horizontal

    return if (orientation == LinearLayout.HORIZONTAL) {
        LinearLayout.LayoutParams(
            if (weight != 0f) 0 else if(sizeX == 0) WRAP_CONTENT else (sizeX * d).toInt(),
            if (align == Align.fill) MATCH_PARENT else if (sizeY == 0) WRAP_CONTENT else (sizeY * d).toInt(),
            weight
        )
    } else {
        LinearLayout.LayoutParams(
            if (align == Align.fill) MATCH_PARENT else if(sizeX == 0) WRAP_CONTENT else (sizeX * d).toInt(),
            if (weight != 0f) 0 else if (sizeY == 0) WRAP_CONTENT else (sizeY * d).toInt(),
            weight
        )
    }.apply {
        this.gravity = when (align) {
            Align.start -> Gravity.TOP or Gravity.START
            Align.center -> Gravity.CENTER
            Align.end -> Gravity.BOTTOM or Gravity.END
            Align.fill -> Gravity.CENTER
        }
        setMargins(
            (d * marginStart).toInt(),
            (d * marginTop).toInt(),
            (d * marginEnd).toInt(),
            (d * marginBottom).toInt()
        )
    }
}

File: ios/Khrysalis/com/lightningkite/khrysalis/views/LinearLayout.actual.swift
//Stub file made with Khrysalis 2 (by Lightning Kite)
import Foundation
import UIKit


//--- LinearLayout.params(Int, Int, Int, Int, Int, Int, AlignPair, Float)
public extension LinearLayout {
    func params(
        _ sizeX: Int32,
        _ sizeY: Int32 = 0,
        _ marginStart: Int32 = 0,
        _ marginEnd: Int32 = 0,
        _ marginTop: Int32 = 0,
        _ marginBottom: Int32 = 0,
        _ gravity: AlignPair = .center,
        _ weight: Float = 0
    ) -> LinearLayout.LayoutParams {
        return LinearLayout.LayoutParams(
            minimumSize: .zero,
            size: CGSize(width: Int(sizeX), height: Int(sizeY)),
            margin: UIEdgeInsets(
                top: CGFloat(marginTop),
                left: CGFloat(marginStart),
                bottom: CGFloat(marginBottom),
                right: CGFloat(marginEnd)
            ),
            gravity: gravity,
            weight: CGFloat(weight)
        )
    }
    func params(
        sizeX: Int32 = 0,
        sizeY: Int32 = 0,
        marginStart: Int32 = 0,
        marginEnd: Int32 = 0,
        marginTop: Int32 = 0,
        marginBottom: Int32 = 0,
        gravity: AlignPair = .center,
        weight: Float = 0
    ) -> LinearLayout.LayoutParams {
        return params(sizeX, sizeY, marginStart, marginEnd, marginTop, marginBottom, gravity, weight)
    }
}
